server: shutdown during startup gets stuck in `unable to log node_restart event` loop

Please answer these questions **for each node in your cluster**:

1. What version of CockroachDB are you using (`cockroach version`)?
`beta-20160101`

2. What operating system and processor architecture are you using?
`Linux c1 4.2.0-34-generic #39-Ubuntu SMP Thu Mar 10 22:13:01 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux`

3. What flags/environment variables did you pass to `cockroach start`?
`start --alsologtostderr --insecure`

Please describe the issue you observed:

1. What did you do?
ctrl-c immediately during restart (i.e. existing `cockroach-data` dir) node.

2. What did you expect to see?
shutdown.

3. What did you see instead?
printed `W160330 18:03:39.350599 server/node.go:664  [node=1] unable to log node_restart event for node 1: the descriptor for the first range is not available via gossip
` in a tight loop (11000x lines in the second it took me to ctrl-c again).
```
david@c1:~$ ./cockroach-beta-20160329.linux-amd64/cockroach start --insecure --alsologtostderr
I160330 18:03:16.501013 cli/start.go:162  [build] beta-20160101 @ 2016/03/30 02:38:57 (go1.6)
I160330 18:03:16.501185 server/context.go:275  1 storage engine initialized
I160330 18:03:16.501209 cli/start.go:176  starting cockroach node
W160330 18:03:16.501230 server/server.go:104  running in insecure mode, this is strongly discouraged. See --insecure.
W160330 18:03:16.501387 gossip/gossip.go:889  not connected to cluster; use --join to specify a connected node
W160330 18:03:16.504116 gossip/gossip.go:889  not connected to cluster; use --join to specify a connected node
I160330 18:03:16.504244 storage/engine/rocksdb.go:120  opening rocksdb instance at "cockroach-data"
I160330 18:03:16.563267 server/node.go:372  initialized store store=1:1 ([]=cockroach-data): {Capacity:10388168704 Available:8588763136 RangeCount:0}
I160330 18:03:16.563783 server/node.go:284  node ID 1 initialized
I160330 18:03:16.563965 storage/stores.go:282  read 22 node addresses from persistent storage
I160330 18:03:16.564281 server/node.go:493  connecting to gossip network to verify cluster ID...
I160330 18:03:16.564353 server/node.go:514  node connected via gossip and verified as part of cluster "9f2db536-3176-4910-9637-bee406320339"
I160330 18:03:16.564674 server/node.go:337  [node=1] Started node with [[]=cockroach-data] engine(s) and attributes []
I160330 18:03:16.564751 server/server.go:359  starting http server at c1:8080
I160330 18:03:16.564885 server/server.go:360  starting grpc/postgres server at c1:26257
build:     beta-20160101 @ 2016/03/30 02:38:57 (go1.6)
admin:     http://c1:8080
sql:       postgresql://root@c1:26257?sslmode=disable
logs:      cockroach-data/logs
store[0]:  path=cockroach-data
^CI160330 18:03:16.874294 cli/start.go:225  initiating graceful shutdown of server
initiating graceful shutdown of server
I160330 18:03:16.874469 stopper.go:352  draining; tasks left:
2      storage/replica.go:1771
1      ts/db.go:100
1      server/node.go:633
W160330 18:03:16.874576 ts/db.go:98  error writing time series data: the descriptor for the first range is not available via gossip
I160330 18:03:16.874605 stopper.go:352  draining; tasks left:
2      storage/replica.go:1771
1      server/node.go:633
I160330 18:03:16.874635 server/updates.go:123  Error reading updates check time: the descriptor for the first range is not available via gossip
W160330 18:03:16.874693 server/node.go:664  [node=1] unable to log node_restart event for node 1: the descriptor for the first range is not available via gossip
```

This is addressed (for another reason) in the last commit of #5512.

On Wed, Mar 30, 2016 at 2:08 PM, David Taylor <notifications@github.com>
wrote:

> Please answer these questions *for each node in your cluster*:
>
>    1.
>
>    What version of CockroachDB are you using (cockroach version)?
>    2.
>
>    What operating system and processor architecture are you using?
>    3.
>
>    What flags/environment variables did you pass to cockroach start?
>
> Please describe the issue you observed:
>
>    1. What did you do? ctrl-c immediately during restart (i.e. existing
>    cockroach-data dir) node.
>    2. What did you expect to see? shutdown.
>    3. What did you see instead? printed W160330 18:03:39.350599
>    server/node.go:664 [node=1] unable to log node_restart event for node 1:
>    the descriptor for the first range is not available via gossip in a
>    tight loop (11000x lines in the second it took me to ctrl-c again).
>
> david@c1:~$ ./cockroach-beta-20160329.linux-amd64/cockroach start --insecure --alsologtostderr
> I160330 18:03:16.501013 cli/start.go:162  [build] beta-20160101 @ 2016/03/30 02:38:57 (go1.6)
> I160330 18:03:16.501185 server/context.go:275  1 storage engine initialized
> I160330 18:03:16.501209 cli/start.go:176  starting cockroach node
> W160330 18:03:16.501230 server/server.go:104  running in insecure mode, this is strongly discouraged. See --insecure.
> W160330 18:03:16.501387 gossip/gossip.go:889  not connected to cluster; use --join to specify a connected node
> W160330 18:03:16.504116 gossip/gossip.go:889  not connected to cluster; use --join to specify a connected node
> I160330 18:03:16.504244 storage/engine/rocksdb.go:120  opening rocksdb instance at "cockroach-data"
> I160330 18:03:16.563267 server/node.go:372  initialized store store=1:1 ([]=cockroach-data): {Capacity:10388168704 Available:8588763136 RangeCount:0}
> I160330 18:03:16.563783 server/node.go:284  node ID 1 initialized
> I160330 18:03:16.563965 storage/stores.go:282  read 22 node addresses from persistent storage
> I160330 18:03:16.564281 server/node.go:493  connecting to gossip network to verify cluster ID...
> I160330 18:03:16.564353 server/node.go:514  node connected via gossip and verified as part of cluster "9f2db536-3176-4910-9637-bee406320339"
> I160330 18:03:16.564674 server/node.go:337  [node=1] Started node with [[]=cockroach-data] engine(s) and attributes []
> I160330 18:03:16.564751 server/server.go:359  starting http server at c1:8080
> I160330 18:03:16.564885 server/server.go:360  starting grpc/postgres server at c1:26257
> build:     beta-20160101 @ 2016/03/30 02:38:57 (go1.6)
> admin:     http://c1:8080
> sql:       postgresql://root@c1:26257?sslmode=disable
> logs:      cockroach-data/logs
> store[0]:  path=cockroach-data
> ^CI160330 18:03:16.874294 cli/start.go:225  initiating graceful shutdown of server
> initiating graceful shutdown of server
> I160330 18:03:16.874469 stopper.go:352  draining; tasks left:
> 2      storage/replica.go:1771
> 1      ts/db.go:100
> 1      server/node.go:633
> W160330 18:03:16.874576 ts/db.go:98  error writing time series data: the descriptor for the first range is not available via gossip
> I160330 18:03:16.874605 stopper.go:352  draining; tasks left:
> 2      storage/replica.go:1771
> 1      server/node.go:633
> I160330 18:03:16.874635 server/updates.go:123  Error reading updates check time: the descriptor for the first range is not available via gossip
> W160330 18:03:16.874693 server/node.go:664  [node=1] unable to log node_restart event for node 1: the descriptor for the first range is not available via gossip
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/issues/5724>
>

--- a/server/node.go
+++ b/server/node.go
@@ -46,6 +46,7 @@ import (
 	"github.com/cockroachdb/cockroach/util/hlc"
 	"github.com/cockroachdb/cockroach/util/log"
 	"github.com/cockroachdb/cockroach/util/metric"
+	"github.com/cockroachdb/cockroach/util/retry"
 	"github.com/cockroachdb/cockroach/util/stop"
 	"github.com/cockroachdb/cockroach/util/timeutil"
 	"github.com/cockroachdb/cockroach/util/tracing"
@@ -648,7 +649,7 @@ func (n *Node) recordJoinEvent() {
 	}
 
 	n.stopper.RunWorker(func() {
-		for {
+		for r := retry.Start(retry.Options{Closer: n.stopper.ShouldStop()}); r.Next(); {
 			if err := n.ctx.DB.Txn(func(txn *client.Txn) *roachpb.Error {
 				return sql.MakeEventLogger(n.ctx.SQLExecutor.LeaseManager).InsertEventRecord(txn,
 					logEventType,
@@ -664,13 +665,6 @@ func (n *Node) recordJoinEvent() {
 				log.Warningc(n.context(context.TODO()), "unable to log %s event for node %d: %s", logEventType, n.Descriptor.NodeID, err)
 			} else {
 				return
-			}
-
-			select {
-			case <-n.stopper.ShouldStop():
-				return
-			default:
-				// Break.
 			}
 		}
 	})

